src/platformio.ts

```diff
@@ -85,21 +98,124 @@ export class PlatformIO {
+                    this.showErrorToAddIncludePath();
+                }
+
                 return;
             }
-            vscode.window.showInformationMessage("Add Include Path to c_cpp_properties.json successfully.");
+
+            if (!silent) {
+                vscode.window.showInformationMessage("Add Include Path to c_cpp_properties.json successfully.");
+            }
         } catch (ex) {
-            this.showErrorToAddIncludePath();
+            if (!silent) {
+                this.showErrorToAddIncludePath();
+            }
         } finally {
-            AppInsightsClient.sendEvent("addIncludePath");
+            if (!silent) {
+                AppInsightsClient.sendEvent("addIncludePath");
+            }
         }
     }
 
     public onDidCloseTerminal(closedTerminal: vscode.Terminal): void {
         this._pioTerminal.onDidCloseTerminal(closedTerminal);
     }
 
+    private configureAutoUpdateIncludes(): void {
+        const cppExtension = vscode.extensions.getExtension("ms-vscode.cpptools");
+        if (cppExtension) {
+            const autoUpdateIncludes = vscode.workspace.getConfiguration("platformio").get<boolean>("autoUpdateIncludes");
+            if (autoUpdateIncludes && !this._autoUpdateIncludesEnabled) {
+                this.buildFolderList();
+                const configFilePath = path.join(vscode.workspace.rootPath, ".vscode", "c_cpp_properties.json");
+                if (!Utility.fileExistsSync(configFilePath)) {
+                    this.addIncludePath(true);
+                }
+
+                // Watch .piolibdeps and lib folders
+                const fileSystemWatcher: vscode.FileSystemWatcher = vscode.workspace.createFileSystemWatcher("**/{.piolibdeps,lib}/**");
+                this._autoUpdateIncludesDisposables.push(fileSystemWatcher);
+                fileSystemWatcher.onDidCreate((e: vscode.Uri) => this.fileSystemDidCreate(e), this, this._autoUpdateIncludesDisposables);
+                fileSystemWatcher.onDidDelete((e: vscode.Uri) => this.fileSystemDidDelete(e), this, this._autoUpdateIncludesDisposables);
+
+                // Watch platformio.ini file
+                const iniFileWatcher: vscode.FileSystemWatcher = vscode.workspace.createFileSystemWatcher("**/platformio.ini");
+                this._autoUpdateIncludesDisposables.push(iniFileWatcher);
+                iniFileWatcher.onDidChange((e: vscode.Uri) => this.addIncludePath(true), this, this._autoUpdateIncludesDisposables);
+                iniFileWatcher.onDidCreate((e: vscode.Uri) => this.addIncludePath(true), this, this._autoUpdateIncludesDisposables);
+            }
+
+            if (!autoUpdateIncludes && this._autoUpdateIncludesEnabled) {
+                this._folderList = undefined;
+                this.disposeAutoUpdateIncludes();
+            }
+
+            this._autoUpdateIncludesEnabled = autoUpdateIncludes;
+        }
+    }
+
+    private fileSystemDidCreate(e: vscode.Uri): void {

```

Comment:
Thanks for your detailed explanation! Yes, we don't care about deleted folders for now, we could get rid of the code than handles deletes.